Solved: 삼성_계산기2

2ssue opened the pull request:
# 삼성 1223 . 계산기2 #1

[문제 링크](https://swexpertacademy.com/main/code/problem/problemDetail.do?contestProbId=AV14nnAaAFACFAYD&categoryId=AV14nnAaAFACFAYD&categoryType=CODE)

| 난이도 | 정답률(\_%) |
| :----: | :---------: |
| D4 | 87.48% |

## 설계
### 고생한 점
- 역시 시간복잡도 계산하기 어렵다 🙄
- 중간에 테스트한다고 출력했던 문구를 제거 안해서 삼성 사이트에서 Fail 뜸.. 이거 찾느라 고생했다
### 복잡도 계산
문자열 길이를 n이라고 함. 
#### 시간복잡도
- 후위 연산으로 변환 할 때 n번 회전하면서, 연산자일 경우 Stack 길이만큼 비교하는 과정이 a번 들어갈 수 있음 (a < n)
- 후위 연산식으로 변환한 식을 계산하는 n번 연산
- 더하면 몇일까요? 😂
#### 공간복잡도
- 입력받은 연산식 char형 길이 n
- 후위 연산식 char형 길이 n
- 후위 연산 계산 Stack
## 풀이
- 중위 연산식 문자열을 후위연산식으로 변환한다. 아래는 변환 방법. 모든 연산식을 돌 때까지 반복한다. 
  - 피연산자일 경우 연산식에 추가
  - 연산자일 경우 연산자 Stack의 가장 위 연산자와 우선순위를 비교
    - (현재 연산자가 가장 위 연산자보다) 우선순위가 낮거나 같은 경우
      - 현재 연산자가 높아질 때까지 Stack의 값을 pop하여 연산식에 추가한 뒤 Stack에 현재 연산자를 넣는다.
    - 우선순위가 높은 경우 
      - Stack에 값을 넣는다.
- 변환한 연산식대로 후위 연산
  - 피연산자는 스택에 쌓고, 연산자일 경우 2개의 피연산자를 꺼내 계산한 값을 넣으면서 계산하는 방식으로 계산하였다.

2ssue committed:
Solved: 삼성_계산기2

changicho commented on problems/samsung/1223/sujeong.java:
```diff
@@ -0,0 +1,120 @@
+        return result;
+    }
+
+    public static char[] makePosteriorOperation(char[] operations) {
```
중위 표기식을 후위 표기식으로 바꾸는 함수네요.

```java
public static char[] makePostfix(char[] infix){
	// do something
}
```

posterior보다 postfix로 명명하는것도 고려해볼만 할것같아요. 또한 내부에 인자로 들어오는 operation 도 중위표기식 임을 한번에 알 수 있습니다.

2ssue replied:
넵 이 이름이 더 좋은 것 같네요 😄 사실 생각이 안나서 사전에 쳤더니 GG..

changicho commented on problems/samsung/1223/sujeong.java:
```diff
@@ -0,0 +1,120 @@
+
+            char operator = operators.peek();
+
+            posteriorOperation.append(comparePriority(operator, operation, operators));
```
스택에 top(peek)이 우선순위가 가장 높게하기 위해서 우선순위 비교를 메소드로 빼주셨는데요. 

operator의 경우 여기서만 사용된다면, operators의 peek임을 바로 알 수 있게 

```java
comparePriority(operators.peek(), operation, operators)
``` 

이렇게 표현하면 코드를 줄일 수 있겠습니다.

2ssue replied:
좋은 방법이네요.. 저도 반복적인 의미가 너무 많기도 해서 마음에 안들었는데 감사합니다 👍

changicho commented on problems/samsung/1223/sujeong.java:
```diff
@@ -0,0 +1,120 @@
+        return posteriorOperation.toString().toCharArray();
+    }
+
+    public static String comparePriority(char operator1, char operator2, Stack<Character> operators) {
```
우선순위 비교와 stack에서 우선순위가 낮은 연산자들을 빼내는 함수인데요.

함수의 명칭에 혼돈이 있어 보입니다.

또한 이 함수는 재귀함수로 구성되어있는데요. 수정님 공부 레포에 꼬리물기최적화에 대한 내용이 있던데 이 재귀함수를 반복문으로 최적화가 가능할까요?

[재귀호출을 반복문으로 바꾸기](https://www.bsidesoft.com/4314)

저는 이 재귀식이 재귀 최적화 (반복문으로)가 가능하다고 보는데요.

1. 재귀호출을 반복문으로 고치려면 이미 재귀함수가 꼬리물기 최적화 형태로 되어있어야 한다.
2. 꼬리물기 최적화가 끝나있는 함수는 return을 결과 변수에 취합한다.

두 조건을 만족하기 때문입니다.

2ssue replied:
넵 감사합니다! 덕분에 다시 tail Recursion에 대해서 학습했어요..ㅎㅎ
호출이 많지 않더라도 끝을 아는 재귀는 자제해보도록 하겠습니다🤦‍♂️

2ssue committed:
Refactor: 리뷰 의견을 듣고 리팩토링

- 재귀로 연산자를 빼던 부분 반복문으로 변경
  (지금보니 왜 재귀로 짰는지 모르겠다. the love..)
- 의미 전달이 잘 안되는 변수명 변경
